hybrid/compartment_examples/inter_comp_call/secure-try_deref/main.c

```diff
@@ -0,0 +1,203 @@
+	void *__capability comp_fn;
+};
+
+// ASM offsets, included here for validation
+#include "main.h"
+
+static_assert(COMP_SIZE == sizeof(struct comp), "Invalid `COMP_SIZE` provided");
+static_assert(COMP_OFFSET_STK_ADDR == offsetof(struct comp, stack_addr),
+			  "Invalid `COMP_OFFSET_STK_ADDR` provided.");
+static_assert(COMP_OFFSET_STK_LEN == offsetof(struct comp, stack_len),
+			  "Invalid `COMP_OFFSET_STK_LEN` provided.");
+static_assert(COMP_OFFSET_DDC == offsetof(struct comp, ddc), "Invalid `COMP_OFFSET_DDC` provided.");
+static_assert(COMP_OFFSET_PCC == offsetof(struct comp, comp_fn),
+			  "Invalid `COMP_OFFSET_PCC` provided.");
+
+struct comp comps[COMP_COUNT];
+
+/*******************************************************************************
+ * Privileged Functions
+ ******************************************************************************/
+
+/* Create and save required capabilities. Currently, this means:
+ * - PCC of `switch_compartment`
+ * - DDC of `switch_compartment`
+ * - capability to allow compartments to call `switch_compartment` via `lpdblr`
+ */
+void init_comps()
+{
+	void *__capability switch_cap = (void *__capability) switch_compartment;
+	switch_cap = cheri_bounds_set(switch_cap, 80 * 4);
+	switcher_caps[1] = switch_cap;
+
+	void *__capability comps_addr = (void *__capability) &comps;
+	comps_addr = cheri_bounds_set(comps_addr, COMP_COUNT * COMP_SIZE);
+	switcher_caps[0] = comps_addr;
+
+	switcher_call = (void *__capability) switcher_caps;
+	// Seal this capability to be only used via a `lpb` type call
+	asm("seal %w0, %w0, lpb" : "+r"(switcher_call) :);
+}
+
+void add_comp(uint8_t *_start_addr, void (*_comp_fn)(), void *_comp_fn_end)
+{
+	assert(id < COMP_COUNT);
+	struct comp new_comp;
+	new_comp.id = id;
+
+	new_comp.compartment_start = (void *) _start_addr;
+	new_comp.stack_addr = (void *) (_start_addr + comp_stack_size);
+	new_comp.stack_len = comp_stack_size;
+	new_comp.heap_len = total_comp_size - comp_stack_size;
+
+	// Ensure 16-byte alignment throught the compartment bounds
+	assert(((uintptr_t) new_comp.compartment_start) % 16 == 0);
+	assert(((uintptr_t) new_comp.stack_addr) % 16 == 0);
+	assert(total_comp_size % 16 == 0);
+
+	// When creating a compartment, store a local copy of the capability which
+	// will allow us to call `switch_compartment` in the heap of the compartment.
+	void *heap_top = (void *) (_start_addr + total_comp_size - sizeof(void *__capability));
+	memcpy(heap_top, &switcher_call, sizeof(void *__capability));
+
+	void *__capability comp_ddc = (void *__capability) _start_addr;
+	comp_ddc = cheri_bounds_set(comp_ddc, total_comp_size);
+	new_comp.ddc = comp_ddc;
+
+	// Set up a capability pointing to the function we want to call within the
+	// compartment. This will be loaded as the PCC when the function is called.
+	void *__capability comp_fn = (void *__capability) _comp_fn;
+	// 40 is arbitary; meant to be the size of the executable function within
+	// compartment
```

Comment:
I think you addressed that with precise pointer arithmetic.

Reply:
Ah, I forgot to update this comment. I'm unsure if we can amend this PR as is, after merging, or should this be done as part of another PR? @ltratt

Reply:
This is merged, so it can be done in another PR. I'm fine with small PRs that fix such issues!